pkg/metrics/diskstats_linux.go

```diff
@@ -0,0 +1,173 @@
+/*
+Copyright 2018 The OpenEBS Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metrics
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"os"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const (
+	diskSubsystem  = ""
+	diskSectorSize = 512
+)
+
+var (
+	previousDiskStats = map[string][]string{}
+	ignoredDevices    = "^(ram|loop|fd|(h|v|xv)d[a-z]|nvme\\d+n\\d+p)\\d*$"
+)
+
+type typedFactorDesc struct {
+	desc      *prometheus.Desc
+	valueType prometheus.ValueType
+	factor    float64
+}
+
+type diskstatsCollector struct {
+	ignoredDevicesPattern *regexp.Regexp
+	descs                 []typedFactorDesc
+}
+
+func init() {
+	getDiskStats()
+	registerCollector("diskstats", defaultEnabled, NewDiskstatsCollector)
+}
+
+func (d *typedFactorDesc) mustNewConstMetric(value float64, labels ...string) prometheus.Metric {
+	if d.factor != 0 {
+		value *= d.factor
+	}
+	return prometheus.MustNewConstMetric(d.desc, d.valueType, value, labels...)
+}
+
+func (c *diskstatsCollector) Update(ch chan<- prometheus.Metric) error {
+	procDiskStats := procFilePath("diskstats")
+	diskStats, err := getDiskStats()
+	if err != nil {
+		return fmt.Errorf("couldn't get diskstats: %s", err)
+	}
+
+	for dev, stats := range diskStats {
+		if c.ignoredDevicesPattern.MatchString(dev) {
+			continue
+		}
+
+		if len(stats) != len(c.descs) {
+			return fmt.Errorf("invalid line for %s for %s", procDiskStats, dev)
+		}
+
+		for i, value := range stats {
+			v, err := strconv.ParseFloat(value, 64)
+			if err != nil {
+				return fmt.Errorf("invalid value %s in diskstats: %s", value, err)
+			}
+			ch <- c.descs[i].mustNewConstMetric(v, dev)
+		}
+	}
+	return nil
+}
+
+func getDiskStats() (map[string]map[int]string, error) {
+	file, err := os.Open(procFilePath("diskstats"))
+	if err != nil {
+		return nil, err
+	}
+	defer file.Close()
+
+	return parseDiskStats(file)
+}
+
+func parseDiskStats(r io.Reader) (map[string]map[int]string, error) {
+	var (
+		diskStats = map[string]map[int]string{}
+		scanner   = bufio.NewScanner(r)
+	)
+
+	for scanner.Scan() {
+		parts := strings.Fields(scanner.Text())
+		if len(parts) < 4 { // we strip major, minor and dev
+			return nil, fmt.Errorf("invalid line in %s: %s", procFilePath("diskstats"), scanner.Text())
+		}
+		dev := parts[2]
+		disk := parts[3:]
```

Comment:
add details about the parts[2], parts[3:] here.

Reply:
Added